main/src/library/List.flix

```diff
@@ -0,0 +1,579 @@
+    ///
+    def range(b: Int32, e: Int32): List[Int32] = if (b >= e) Nil else b :: range(b + 1, e)
+
+    ///
+    /// Returns a list with the element `x` repeated `n` times.
+    ///
+    def repeat[a](x: a, n: Int32): List[a] = if (n <= 0) Nil else x :: repeat(x, n - 1)
+
+    ///
+    /// Alias for `scanLeft`.
+    ///
+    def scan[a,b](f: (b, a) -> b, s: b, xs: List[a]): List[b] = scanLeft(f, s, xs)
+
+    ///
+    /// Accumulates the result of applying `f` to `xs` going left to right.
+    /// That is, the result is of the form: `s :: f(s, x1) :: f(f(s, x1), x2)  ...`.
+    ///
+    def scanLeft[a,b](f: (b, a) -> b, s: b, xs: List[a]): List[b] = s :: scanLeftHelper(f, s, xs)
+
+    ///
+    /// Helper function for `scanLeft`.
+    /// Accumulates the result of applying `f` to `xs` going left to right without including `s` at the beginning.
+    /// That is, scanLeft(f, s, xs) = s :: scanLeftHelper(f, s, xs).
+    ///
+    def scanLeftHelper[a,b](f: (b, a) -> b, s: b, xs: List[a]): List[b] = match xs with {
+        case Nil => Nil
+        case x :: rs => f(s, x) :: scanLeftHelper(f, f(s, x), rs)
+    }
+
+    ///
+    /// Accumulates the result of applying `f` to `xs` going right to left.
+    /// That is, the result is of the form: `... f(xn-1, f(xn, s)) :: f(xn, s) :: s`.
+    ///
+    def scanRight[a,b](f: (a, b) -> b, s: b, xs: List[a]): List[b] = match xs with {
+        case Nil => s :: Nil
+        case x :: rs => match scanRight(f, s, rs) with {
+            case y :: ys => f(x, y) :: y :: ys
+        }
+    }
+
+    //
+    // ## List Transformation
+    //
+
+    ///
+    /// Returns the result of applying `f` to every element in `xs`.
+    /// That is, the result is of the form: `f(x1) :: f(x2) :: ...`.
+    ///
+    def map[a,b](f: a -> b, xs: List[a]): List[b] = match xs with {
+        case Nil => Nil
+        case x :: rs => f(x) :: map(f, rs)
+    }
+
+    ///
+    /// Returns the result of applying `f` to every element in `xs` along with that element's index.
+    /// That is, the result is of the form: `f(x1, 0) :: f(x2, 1) :: ...`.
+    ///
+    def mapWithIndex[a,b](f: (a, Int32) -> b, xs: List[a]): List[b] = mapWithIndexHelper(f, xs, 0)
+
+    ///
+    /// Helper function for `mapWithIndex`.
+    /// Returns the result of applying `f` to every element in `xs` along with that element's index + `i`.
+    /// That is, the result is of the form: `f(x1, i) :: f(x2, i+1) :: ...`.
+    /// mapWithIndex(f, xs) = mapWithIndexHelper(f, xs, 0).
+    ///
+    def mapWithIndexHelper[a,b](f: (a, Int32) -> b, xs: List[a], i: Int32): List[b] = match xs with {
+        case Nil => Nil
+        case x :: rs => f(x, i) :: mapWithIndexHelper(f, rs, i+1)
+    }
+
+    ///
+    /// Returns the result of applying `f` to every element in `xs` and concatenating the results.
+    ///
+    def flatMap[a,b](f: a -> List[b], xs: List[a]): List[b] = match xs with {
```

Comment:
I believe this is O(n^2) if `xs` is length `n` and `f` outputs a list of length `n`, assuming that appending a list of length `n` onto another list takes O(n) time.  My analysis might be wrong though, I'm unsure how long appending takes in Flix.